odoo/odoo
# [FIX] product: add explicit copy args to field 'property_product_pricelist' so we can use default pricelists in B2C mode

Impacted versions: 
- Tested on Odoo14 Community and Enterprise

Steps to reproduce:

- Enable debug mode, go through Settings/Website and enable B2C mode (Free Sign Up)
- Go through Settings/Sales and enable Pricelists (any mode)
- Save
- Create pricelists or use demo data
- Go through 'Contact' menu, apply filter for 'Archivied' records and make a search for 'Portal User Template'
- Open Portal User Template, go to TAB "sale and purchase", edit, select pricelist to assign and Save.
- Logout, login on website and create a new portal user

Current behavior:

- The default pricelist assigned to the newly created partner will not be assigned by Portal User Template, it will be instead the first pricelist in the sequence (the product.pricelist model ordered by default on "id desc")

Expected behavior:

- If we set the field  'property_product_pricelist' (basically the pricelist field) in the Portal User Template, it should be used as the default pricelist for Portal Users

Description of the issue/feature this PR addresses:

- The unexpected behaviour happens because [signup](https://github.com/odoo/odoo/blob/fb7dbfb03f4e7c48d199c98fc852e037d3127d16/addons/auth_signup/models/res_users.py#L103)  method calls > [signup_create_user()](https://github.com/odoo/odoo/blob/fb7dbfb03f4e7c48d199c98fc852e037d3127d16/addons/auth_signup/models/res_users.py#L112) > [_create_user_from_template()](https://github.com/odoo/odoo/blob/fb7dbfb03f4e7c48d199c98fc852e037d3127d16/addons/auth_signup/models/res_users.py#L134) , it will return a [copy of the template merged with user provided values](https://github.com/odoo/odoo/blob/fb7dbfb03f4e7c48d199c98fc852e037d3127d16/addons/auth_signup/models/res_users.py#L149) to create new contact. <br/><br/>The 'copy' will not afftect the field 'property_product_pricelist' though, because it's a computed field so the copy parameter must be explicit (found answer in [orm api fields definition 14.0](https://www.odoo.com/documentation/14.0/developer/reference/addons/orm.html#odoo.fields.Field)).<br/>

> copy ([bool](https://docs.python.org/3/library/functions.html#bool)) – whether the field value should be copied when the record is duplicated (default: True for normal fields, False for one2many and computed fields, including property fields and related fields)

After PR: 
- I tested with copy=True, and it seems to resolve the issue for Portal Users signup.

Additional notes:
-  I have also tested the B2B mode, surprisingly it seems to not be solved by this, but the B2C signup will benefit from this PR.



---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
